lib/block_view_lockups.go

```diff
@@ -818,109 +1174,541 @@ func (bav *UtxoView) _connectCoinLockup(
+				//                                                                                    ^  ^
+				//                                                                               right overhang
+				// We will break any overhang off into its own separate locked balance entry.
+
+				// Check for right overhang by the existing locked balance entry
+				if existingLockedBalanceEntry.VestingEndTimestampNanoSecs >
+					proposedLockedBalanceEntry.VestingEndTimestampNanoSecs {
+					splitLockedBalanceEntry, remainingLockedBalanceEntry, err := SplitVestedLockedBalanceEntry(
+						existingLockedBalanceEntry,
+						proposedLockedBalanceEntry.VestingEndTimestampNanoSecs+1,
+						existingLockedBalanceEntry.VestingEndTimestampNanoSecs)
+					if err != nil {
+						return 0, 0, nil,
+							errors.Wrap(err, "_connectCoinLockup failed to compute vested split")
+					}
+
+					// Set the splitLockedBalanceEntry into the view.
+					bav._setLockedBalanceEntry(splitLockedBalanceEntry)
+
+					// We update the existingLockedBalanceEntry as broke the right overhanging portion off.
+					existingLockedBalanceEntry = remainingLockedBalanceEntry
+				}
+
+				// Check for right overhang by the proposed locked balance entry
+				if proposedLockedBalanceEntry.VestingEndTimestampNanoSecs >
+					existingLockedBalanceEntry.VestingEndTimestampNanoSecs {
+					// NOTE: This case is particularly interesting as there's two situations in
+					// which we might find ourselves.
+					//
+					// First case:
+					// proposed:   <----------------------------------------->
+					// existing:   <------------------>
+					//                                 ^                     ^ Overhang
+					// Second case:
+					// proposed:   <----------------------------------------->
+					// existing:   <------------------>            <--------->
+					//                                 ^          ^ Overhang
+					// In the second case there exists another conflicting LockedBalanceEntry
+					// sometime in the future that we must be aware of. Also note that
+					// the first case is only possible in the very last iteration.
+					//
+					// To account for this, we split from the proposed LockedBalanceEntry
+					// and combine the left overlapping portion. This leaves us with two
+					// remaining cases:
+					// First case:
+					// proposed:                       <--------------------->
+					// existing:
+					//
+					// Second case:
+					// proposed:                       <--------------------->
+					// existing:                                   <--------->
+					//
+					// The second case is fine to leave as it will be taken care of in the
+					// subsequent iteration. However, we make a special note to capture the
+					// remaining LockedBalanceEntry present in the first case.
+
+					// We check if there's another locked balance entry sometime in the future.
+					splitLockedBalanceEntry, remainingLockedBalanceEntry, err := SplitVestedLockedBalanceEntry(
```

Comment:
This is super tricky. I think it works only because we've cut off everything on the left at this point, and so we're 100% sure that the existing entry wholly overlaps with the proposed one AND the proposed one doesn't overhang the existing entry. Only if you have both of those, THEN you can consolidate the left parts of the thing. I think it wouldn't hurt to make it clear that all of the previous chopping PLUS the fact that existing entries can't overlap with each other make this work.